module/zfs/dmu_objset.c

```diff
@@ -1610,6 +1610,22 @@ dmu_objset_write_done(zio_t *zio, arc_buf_t *abuf, void *arg)
 
 		(void) dsl_dataset_block_kill(ds, bp_orig, tx, B_TRUE);
 		dsl_dataset_block_born(ds, bp, tx);
+
+		if (ds != NULL) {
+			for (spa_feature_t f = 0; f < SPA_FEATURES; f++) {
+				if (zfeature_active(f,
+				    ds->ds_feature_activation[f])) {
+					if (zfeature_active(f,
+					    ds->ds_feature[f]))
+						continue;
+					dsl_dataset_activate_feature(
+					    ds->ds_object, f,
+					    ds->ds_feature_activation[f], tx);
+					ds->ds_feature[f] =
+					    ds->ds_feature_activation[f];
+				}
+			}
+		}
```

Comment:
I am wondering if this should just be moved into `dal_dataset_block_born()`? At the beginning of that function, it checks if `ds == NULL` and returns if so. This would allow you to avoid doing the `ds != NULL` check here and also keeps `zfeature_active()` as a static function in `dsl_dataset.c`.

Reply:
If we move it there, then it gets called by `dbuf_write_done()` which results in panics like:
```
[  +1.191461] VERIFY3(0 == zap_add(mos, dsobj, spa_feature_table[f].fi_guid, sizeof (zero), 1, &zero, tx)) failed (0 == 17)
[  +0.000722] PANIC at dsl_dataset.c:1115:dsl_dataset_activate_feature()
[  +0.000416] Showing stack for process 2680188
[  +0.000002] CPU: 0 PID: 2680188 Comm: z_wr_int_1 Tainted: P           OE     5.15.61-1-lts #1 c112f9744322a087919db3f6cc619e1f1b1cc16b
[  +0.000003] Hardware name: Gigabyte Technology Co., Ltd. Z390 AORUS ELITE/Z390 AORUS ELITE-CF, BIOS F10 11/05/2021
[  +0.000000] Call Trace:
[  +0.000002]  <TASK>
[  +0.000002]  dump_stack_lvl+0x45/0x5b
[  +0.000006]  spl_panic+0xf0/0x108 [spl f2ecb2706f2af11ac25922bdf0adcf84e1a6928a]
[  +0.000016]  ? queued_spin_unlock+0x5/0x10 [zfs 5adb61136bb00385c870db5ec31532ff702723cc]
[  +0.000159]  ? __raw_spin_unlock+0x5/0x10 [zfs 5adb61136bb00385c870db5ec31532ff702723cc]
[  +0.000146]  ? zap_add_impl+0x91/0x190 [zfs 5adb61136bb00385c870db5ec31532ff702723cc]
[  +0.000178]  dsl_dataset_activate_feature+0x1b4/0x1f0 [zfs 5adb61136bb00385c870db5ec31532ff702723cc]
[  +0.000168]  dsl_dataset_block_born+0x32e/0x690 [zfs 5adb61136bb00385c870db5ec31532ff702723cc]
[  +0.000195]  dbuf_write_done+0x412/0x710 [zfs 5adb61136bb00385c870db5ec31532ff702723cc]
[  +0.000152]  arc_write_done+0x9f/0x710 [zfs 5adb61136bb00385c870db5ec31532ff702723cc]
[  +0.000158]  zio_done+0x3dc/0x18d0 [zfs 5adb61136bb00385c870db5ec31532ff702723cc]
[  +0.000160]  ? kfree+0x221/0x250
[  +0.000005]  zio_execute+0xe7/0x220 [zfs 5adb61136bb00385c870db5ec31532ff702723cc]
[  +0.000161]  taskq_thread+0x1f7/0x430 [spl f2ecb2706f2af11ac25922bdf0adcf84e1a6928a]
[  +0.000012]  ? wake_up_q+0x90/0x90
[  +0.000004]  ? zio_execute_stack_check.constprop.0+0x10/0x10 [zfs 5adb61136bb00385c870db5ec31532ff702723cc]
[  +0.000161]  ? taskq_thread_spawn+0x50/0x50 [spl f2ecb2706f2af11ac25922bdf0adcf84e1a6928a]
[  +0.000011]  kthread+0x115/0x140
[  +0.000002]  ? set_kthread_struct+0x50/0x50
[  +0.000002]  ret_from_fork+0x1f/0x30
[  +0.000004]  </TASK>
```

It is unclear though to me why this happens.

Reply:
Interesting. Where exactly did you put this inside of `dsl_dataset_block_born()`? I wasn't clear on my idea here and sorry about that. I was proposing it be moved to the end of `dsl_dataset_block_born()`. My thought was that code was being run anyways, so if we have gotten to the end of `dsl_dataset_block_born()` with no issues, I would think it would line up just the same as where you put the `dsl_dataset_activate_feature()` currently after calling `dsl_dataset_block_born()`? Am I wrong on that though?

Reply:
I placed it at the end of dsl_dataset_block_born() function. It seemed reasonable to me, too.

Reply:
That is super strange. Do you have a test case that reliably hits that `VERIFY0`? If so, I can help out and take a look.

Reply:
Alright, yeah I was able to duplicate with `zloop.sh` pretty fast lol... That is so strange... I will look into this more myself. I really don't know what the difference is off hand, but I will have to stare at some code to figure it out.

Reply:
Okay, wait... Now I am seeing it. So yeah maybe where you have this make sense then. It will take a thorough combing over everything, but I see in `dbuf.c` how this happens. You probably have this in the right spot then. What I am not certain of entirely is why some of the DSL features are not set at that point. However, where you have this is probably right.

Reply:
I think it would make more sense to have this code in dsl_dataset_sync_done().  Sorry I didn't put it there to begin with!